### PR TITLE
Fork branch cetko refactor

### PR DESCRIPTION
**Pull request by sasakalaba**

### CODE REVIEW

File: tests/helpers.py [hunk 1]
```diff
@@ -14,6 +14,10 @@
 from flask.ext.stormpath import StormpathManager
 from stormpath.client import Client
 
+# FIXME: setup a better way to load environment variables
+environ['STORMPATH_API_KEY_ID'] = '15O7VLV850461TYBRFP91KRR4'
+environ['STORMPATH_API_KEY_SECRET'] = '8Ao/UesWQVhVkE7LL7ZVApHWn/r0cygrrHaruh75ipk'
+
```
Comment (sasakalaba) [inline tests/helpers.py:20]: Temporary fix dok ne rijesim bitnije stvari. Ovo cu maknut cim prije.

File: flask_stormpath/__init__.py [hunk 2]
```diff
@@ -250,6 +250,21 @@ def check_settings(self, config):
...
+        ):
+            raise ConfigurationError(
+                'You must define your Stormpath credentials.')
+
+        if not all([
+            config['stormpath']['application']['href'],
+            config['stormpath']['application']['name']
+        ]):
+            raise ConfigurationError(
+                'You must define your Stormpath application.')
```
Comment (sasakalaba) [inline flask_stormpath/__init__.py:271]: Ovo moram maknuti jer smo rekli na reviewu da se ovo vadi iz ValidateClientConfigStrategy.

File: flask_stormpath/views.py [hunk 3]
```diff
@@ -177,6 +179,8 @@ def login():
     template that is used to render this page can all be controlled via
     Flask-Stormpath settings.
     """
+    # FIXME ASAP: cannot initialize StormpathForm properly
+    form = LoginForm()
     form = LoginForm(csrf_enabled=False)
```
Comment (sasakalaba) [inline flask_stormpath/views.py:228]: Ovog me sram, but alas nisam vise znao kaj da radim sa tom formom, pa sam stavio ovo da mogu dalje view-ove raditi. Probleme sa formama cemo proci u detalje u srijedu.

File: tests/test_views.py [hunk 4]
```diff
@@ -7,15 +7,30 @@
...
+    """
+    def __init__(self, app):
+        self.app = app
+
+    def __call__(self, environ, start_response):
+        environ['HTTP_ACCEPT'] = ('text/html,application/xhtml+xml,' +
+            'application/xml;')
+        return self.app(environ, start_response)
+
+
```
Comment (sasakalaba) [inline tests/test_views.py:37]: Ovako sam ubacio Accept header u request buduci da ga `request_wants_json()` provjerava u view-ovima.

File: flask_stormpath/views.py [hunk 5]
```diff
@@ -72,26 +72,35 @@ def register():
...
+
+        # Check if optional fields are enabled. If not, set them to 'Anonymous'
+        data = form.data
+        for field in optional_fields:
+            if field not in data:
+                if not field_properties[field]['enabled']:
+                    data[field] = 'Anonymous'
+            else:
+                if not data[field] and not field_properties[field]['required']:
+                    data[field] = 'Anonymous'
```
Comment (sasakalaba) [inline flask_stormpath/views.py]: Provjeravam opcionalne form fieldove i u slucaju da su enabled i nemaju vrijednost, postavljam na 'Anonymous', buduci da se proslijediti na `User.create(**data)` te moraju imati vrijednost razlicitu od None ili blank.

File: flask_stormpath/__init__.py [hunk 6]
```diff
@@ -122,6 +135,181 @@ def init_app(self, app):
...
+            )
+
+        ecfrcs = EnrichClientFromRemoteConfigStrategy(
+            client_factory=lambda client: self.client)
+        ecfrcs.process(self.app.config['stormpath'].store)
+        eifrcs = EnrichIntegrationFromRemoteConfigStrategy(
+            client_factory=lambda client: self.client)
+        eifrcs.process(self.app.config['stormpath'].store)
+
+        self.application = self.client.applications.get(
```
Comment (avojnovicDk) [inline flask_stormpath/__init__.py:243]: This Stormpath application.

File: flask_stormpath/__init__.py [hunk 7]
```diff
@@ -97,11 +109,11 @@ def init_app(self, app):
         """
         # Initialize all of the Flask-Stormpath configuration variables and
         # settings.
-        init_settings(app.config)
+        self.init_settings(app.config)
 
         # Check our user defined settings to ensure Flask-Stormpath is properly
         # configured.
-        check_settings(app.config)
+        self.check_settings(app.config)
```
Comment (avojnovicDk) [inline flask_stormpath/__init__.py]: `self.init_stormpath(app.config)` koji inicijalizira Stormpathov `client` i `application` (iz `init_settings` funkcije)

File: flask_stormpath/__init__.py [hunk 8]
```diff
@@ -97,11 +109,11 @@ def init_app(self, app):
...
```
Comment (avojnovicDk) [inline flask_stormpath/__init__.py:110]: We don't need this because we validated settings in `init_settings`.

File: flask_stormpath/views.py [hunk 9]
```diff
@@ -138,10 +149,12 @@ def register():
                 redirect_url = current_app.config[
                     'stormpath']['web']['register']['nextUri']
                 if not redirect_url:
-                    redirect_url = current_app.config[
+                    login_redirect = current_app.config[
```
Comment (avojnovicDk) [inline flask_stormpath/views.py]: web_config.get('register', web_config.get('login', '/'))

File: flask_stormpath/views.py [hunk 10]
```diff
@@ -65,68 +64,51 @@ def register():
...
+        if not form.validate_on_submit():
+        # If form.data is not valid, flash error messages.
+            for field_error in form.errors.keys():
+                flash(form.errors[field_error][0])
+
+        else:
+            # We'll just set the field values to 'Anonymous' if the user
+            # has explicitly said they don't want to collect those fields.
+            for field in ['given_name', 'surname']:
+                if field not in data or not data[field]:
```
Comment (avojnovicDk) [inline flask_stormpath/views.py]: `if not data.get(field)`:

File: flask_stormpath/__init__.py [hunk 11]
```diff
@@ -267,23 +264,31 @@ def check_settings(self, config):
                 'You must define your Stormpath application.')
 
         if config['STORMPATH_ENABLE_GOOGLE']:
-            google_config = config['STORMPATH_SOCIAL'].get('GOOGLE')
+            if 'STORMPATH_SOCIAL' in config:
+                google_config = config['STORMPATH_SOCIAL'].get('GOOGLE')
+            else:
+                google_config = None
```
Comment (avojnovicDk) [inline flask_stormpath/__init__.py:276]: `google_config = config.get('STORMPATH_SOCIAL', {}).get('GOOGLE', {})`

File: flask_stormpath/__init__.py [hunk 12]
```diff
@@ -267,23 +264,31 @@ def check_settings(self, config):
...
 
             if not google_config or not all([
                 google_config.get('client_id'),
                 google_config.get('client_secret'),
             ]):
```
Comment (avojnovicDk) [inline flask_stormpath/__init__.py:281]: No need for `not google_config` if `google_config` is a dict.

File: flask_stormpath/models.py [hunk 13]
```diff
@@ -73,14 +74,24 @@ def delete(self):
...
-        return json.dumps({'account': {
-            key: getattr(self, key, None) for key in writable_attrs}})
+            'full_name'
+        )
+
+        json_data = {'account': {}}
+        for key in writable_attrs:
+            attr = getattr(self, key, None)
+            json_data['account'][key] = (
+                attr if not isinstance(attr, datetime) else attr.isoformat())
```
Comment (avojnovicDk) [inline flask_stormpath/models.py]: `json_data = {'account': {attr: getattr(self, attr, None) for attr in attrs}}`

File: flask_stormpath/models.py [hunk 14]
```diff
@@ -73,14 +74,24 @@ def delete(self):
...
+        return json.dumps(json_data)
```
Comment (avojnovicDk) [inline flask_stormpath/models.py]: `return json.dumps(json_data, default=datetime_handler)`

File: flask_stormpath/models.py [hunk 15]
```diff
@@ -73,14 +74,24 @@ def delete(self):
...
             'username',
             'email',
             'given_name',
             'middle_name',
             'surname',
-            'status')
-        return json.dumps({'account': {
-            key: getattr(self, key, None) for key in writable_attrs}})
+            'full_name'
+        )
```
Comment (avojnovicDk) [inline flask_stormpath/models.py:97]: needs better naming - these are not (only) writable attributes

File: flask_stormpath/__init__.py [hunk 16]
```diff
@@ -198,8 +195,8 @@ def init_settings(self, config):
         config.setdefault('REMEMBER_COOKIE_NAME', 'stormpath_token')
 
         for key, value in config.items():
-            if key.startswith(config['stormpath'].STORMPATH_PREFIX) and \
-                    key in config['stormpath']:
+            if (key.startswith(config['stormpath'].STORMPATH_PREFIX) and
+                    key in config['stormpath']):
                 config['stormpath'][key] = value
```
Comment (avojnovicDk) [inline flask_stormpath/__init__.py:205]: This does nothing.

File: flask_stormpath/__init__.py [hunk 17]
```diff
@@ -267,23 +264,31 @@ def check_settings(self, config):
...
-                raise ConfigurationError('You must define your Google app settings.')
+                raise ConfigurationError(
+                    'You must define your Google app settings.')
 
         if config['STORMPATH_ENABLE_FACEBOOK']:
-            facebook_config = config['STORMPATH_SOCIAL'].get('FACEBOOK')
+            if 'STORMPATH_SOCIAL' in config:
+                facebook_config = config['STORMPATH_SOCIAL'].get('FACEBOOK')
+            else:
+                facebook_config = None
```
Comment (avojnovicDk) [inline flask_stormpath/__init__.py:289]: same as with Google

File: flask_stormpath/__init__.py [hunk 18]
```diff
@@ -267,23 +264,31 @@ def check_settings(self, config):
...
 
             if not facebook_config or not all([
                 facebook_config,
                 facebook_config.get('app_id'),
                 facebook_config.get('app_secret'),
             ]):
-                raise ConfigurationError('You must define your Facebook app settings.')
+                raise ConfigurationError(
+                    'You must define your Facebook app settings.')
```
Comment (avojnovicDk) [inline flask_stormpath/__init__.py:297]: same as with Google

File: flask_stormpath/models.py (hunk 14 above)
Comment (avojnovicDk) [inline flask_stormpath/models.py]: http://stackoverflow.com/questions/23285558/datetime-date2014-4-25-is-not-json-serializable-in-django

File: flask_stormpath/views.py [hunk 19]
```diff
@@ -32,22 +28,28 @@
     FACEBOOK = True
```
Comment (avojnovicDk) [inline flask_stormpath/views.py]: Just a reminder - this has changed in the meantime.

File: flask_stormpath/views.py [hunk 20]
```diff
@@ -555,27 +216,331 @@ def google_login():
...
+    both html and json responses.
+    Specialized logic for each view is handled in the process_request method
+    and should be specified on each child class.
+    """
+
+    def __init__(self, config, *args, **kwargs):
+        self.config = config
+        self.form = (
+            StormpathForm.specialize_form(config['form'])()
+            if config else None)
```
Comment (avojnovicDk) [inline flask_stormpath/views.py]: Check if there is need for this (any case when `self.form` needs to be `None`?)

File: flask_stormpath/views.py [hunk 21]
```diff
@@ -555,27 +216,331 @@ def google_login():
...
+        best = request.accept_mimetypes.best_match(current_app.config[
+            'stormpath']['web']['produces'])
+        if best is None and current_app.config['stormpath']['web']['produces']:
+            best = current_app.config['stormpath']['web']['produces'][0]
+        return best == 'application/json'
+
+    def process_request(self):
+        """ Custom logic specialized for each view. Must be implemented in
+            the subclass. """
+        raise StormpathForm('You must implement process_request on your view.')
```
Comment (avojnovicDk) [inline flask_stormpath/views.py]: `raise NotImplementedError('Subclasses must implement this method.')`

File: flask_stormpath/views.py [hunk 22]
```diff
@@ -555,27 +216,331 @@ def google_login():
...
+        # If we're able to successfully retrieve the user's account,
+        # we'll log the user in (creating a secure session using
+        # Flask-Login), then redirect the user to the ?next=<url>
+        # query parameter, or the Stormpath login nextUri setting.
+        login_user(account, remember=True)
+
+        if self.request_wants_json():
+            return self.make_stormpath_response(data=current_user.to_json())
+
+        return redirect(request.args.get('next') or self.config['nextUri'])
```
Comment (avojnovicDk) [inline flask_stormpath/views.py]: use `if not` as in `RegisterView`.

**Pull request closed without merging**